lib/elastic_apm/transport/connection/state.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module ElasticAPM
+  module Transport
+    # @api private
+    class Connection
+      # @api private
+      class State
+        STATES = {
+          disconnected: 0,
+          connecting: 1,
+          connected: 2
+        }.freeze
+
+        def initialize(value: STATES[:disconnected])
+          @atom = Concurrent::AtomicFixnum.new(value)
+        end
+
+        def value=(value)
+          @atom.value = value
+        end
+
+        def value
+          @atom.value
+        end
+
+        def hold
+          @atom.update do
+            yield self
+            nil
+          end
+        end
+
+        def disconnected!
+          self.value = STATES[:disconnected]
+        end
+
+        def disconnected?
+          value == STATES[:disconnected]
+        end
+
+        def connecting!
+          self.value = STATES[:connecting]
+        end
+
+        def connecting?
+          value == STATES[:connecting]
```

Comment:
I don't think the `connecting?` state is used anywhere.

Reply:
https://github.com/elastic/apm-agent-ruby/pull/372/files#diff-ab7ddf12ccb784736e897ed2ec4a71d2R143 is here! The `?` version might not be used but there for consistency.